### PR TITLE
🚸 Improve search

### PR DESCRIPTION
**Pull request by sunnyosun**
Key improvements:
- prioritize `startswith` and isolated phrases (e.g. "naive B cell", "B cell, ..." over "club cell" when searching "b cell")
- sort by shorter names first

Note: `centrocyte` appears in the "b cell" search because there's a perfect match of "B cell" in the description, same for "t cell" results.

LaminDB Before|LaminDB After |Hub
---|---|---
Screenshot 2024-11-07 at 12 24 25|Screenshot 2024-11-07 at 12 42 48|Screenshot 2024-11-07 at 12 21 23
Screenshot 2024-11-07 at 12 24 31|Screenshot 2024-11-07 at 12 43 10|Screenshot 2024-11-07 at 12 22 12
Screenshot 2024-11-07 at 12 24 17|Screenshot 2024-11-07 at 12 43 18|Screenshot 2024-11-07 at 12 22 34
Screenshot 2024-11-07 at 13 10 44|Screenshot 2024-11-07 at 13 10 15|Screenshot 2024-11-07 at 13 11 40
Screenshot 2024-11-07 at 16 06 53|Screenshot 2024-11-07 at 21 43 05|Screenshot 2024-11-07 at 16 11 32
Screenshot 2024-11-07 at 16 09 37|Screenshot 2024-11-07 at 21 43 25|Screenshot 2024-11-07 at 16 11 58

### CODE REVIEW

**Comment by Koncopd** [comment]
@sunnyosun @falexwolf here is the benchmark for this PR https://github.com/laminlabs/lamindb/pull/2141

**Review by Koncopd — APPROVED** [review]

**Comment by falexwolf** [comment]
These are great improvements!

I wish there was good way in laminhub to document such changes but I guess there isn't for now.

**Comment by falexwolf** [comment]
@fredericenard, please also take a look here.

And then we'll discuss in the benchmarking PR how we proceed with organizing code across lamindb, bionty and laminhub.

**Review by falexwolf — APPROVED** [review]

**Review by falexwolf — CHANGES_REQUESTED** [review]
One thing that's totally not covered is longer search phrases along the lines of "CD8-positie cytokine T cell"

So, I believe we need a benchmarking case for this.

**Comment by Koncopd** [comment]
Added more searches to the benchmark. 
`test_search_synonyms`, commented out in the benchmark.

**Comment by sunnyosun** [comment]
Fixed synonyms. The only thing is that it's getting a bit slower now ~0.8s per search (before 0.2-0.3s) because of all the different layers.

File: lamindb/_record.py [hunk 1]
```diff
@@ -218,8 +221,30 @@ def _search(
                     ) from error
             else:
                 fields.append(field)
+    if not name_field:
+        name_field = (
+            fields[0] if not hasattr(registry, "_name_field") else registry._name_field
+        )
+
+    def tokenize_search_string(search_str: str) -> list[str]:
+        # Split the string
```
Comment (Zethson) [inline lamindb/_record.py:230]:
```suggestion
```

Comment (Zethson) [reply lamindb/_record.py:230]: Self documenting code here.

**Review by Zethson — COMMENTED** [review]
This is really cool!

I know that no review of me was requested but I thought I'd leave a couple of comments still. Feel free to ignore them!

1. Would it be possible to add 1-3 sentences of high level motivation and explanation of this new layered algorithm to the docstring or where appropriate, please?
2. We have the nice search benchmark now. Is there a way to add slightly more sophisticated tests to ensure that no changes of this code will lead to regressions of the search performance? @Koncopd benchmarking notebook is great but it's not regularly run, is it?

These things can also be added later if you think that they're useful...

**Comment by Zethson** [comment]
Ohh and can we get rid of this now?
image
🚀

**Comment by falexwolf** [comment]
Yes, removing will be the goal.

image

@Koncopd is making a push to consolidate all 3 search algorithms into one clearly documented and benchmarked solution.

Sunny's PR here has good ideas but it's not suitable for the hub due to performance. So, the hope is Sergei can replicate the UX with server-side code (essentially correctly using postgres plugins). We can still use Sunny's code for dataframes and sqlite; it'll give the same results but just run slower which is OK in that context.

**Pull request closed without merging**